File: go/orm/internal/conformance/entities_test.go
package conformance

import (
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/testsupport"
	"github.com/kosmassp/SimpleOrm/go/orm/sample"
)

// entityTypes are the ten fixture entities of dotnet/samples/SimpleOrm.Sample,
// mirrored one to one in orm/sample (CODING-STANDARD §7).
var entityTypes = []reflect.Type{
	reflect.TypeFor[sample.User](),
	reflect.TypeFor[sample.Role](),
	reflect.TypeFor[sample.UserRole](),
	reflect.TypeFor[sample.Transaction](),
	reflect.TypeFor[sample.TransactionDetail](),
	reflect.TypeFor[sample.UserProfile](),
	reflect.TypeFor[sample.UserTransactionTotal](),
	reflect.TypeFor[sample.MonthlySalesTotal](),
	reflect.TypeFor[sample.DailySales](),
	reflect.TypeFor[sample.UserActivityReport](),
}

// TestEntities_MatchConformanceFiles is the conformance runner for entity
// metadata (CLAUDE.md §9): every fixture entity's export must be
// byte-identical (modulo line-ending normalization) to
// conformance/entities/<snake_case name>.json — the file every port
// reproduces.
func TestEntities_MatchConformanceFiles(t *testing.T) {
	loader := metadata.NewLoader(nil)
	dir := testsupport.ConformanceFolder(t, "entities")
	covered := map[string]bool{}

	for _, entityType := range entityTypes {
		entityType := entityType
		fileName := core.ToSnakeCase(entityType.Name()) + ".json"
		covered[fileName] = true

		t.Run(entityType.Name(), func(t *testing.T) {
			m, err := loader.Load(entityType)
			if err != nil {
				t.Fatalf("load %s: %v", entityType.Name(), err)
			}

			actual := testsupport.NormalizeNewlines(metadata.Export(m))
			expectedBytes, err := os.ReadFile(filepath.Join(dir, fileName))
			if err != nil {
				t.Fatalf("read conformance/entities/%s: %v", fileName, err)
			}
			expected := strings.TrimRight(testsupport.NormalizeNewlines(string(expectedBytes)), "\n")

			if actual != expected {
				t.Errorf("export mismatch for %s (conformance/entities/%s):\n--- expected ---\n%s\n--- actual ---\n%s",
					entityType.Name(), fileName, expected, actual)
			}
		})
	}

	t.Run("every conformance file is covered by a case", func(t *testing.T) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			t.Fatalf("read conformance/entities: %v", err)
		}
		var names []string
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".json") {
				names = append(names, e.Name())
			}
		}
		sort.Strings(names)
		for _, name := range names {
			if !covered[name] {
				t.Errorf("conformance/entities/%s has no matching entity type in this runner", name)
			}
		}
	})
}

File: php/tests/Conformance/EntitiesTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Conformance;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Metadata\EntityMapJson;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Tests\Sample\Models\DailySales;
use SimpleOrm\Tests\Sample\Models\MonthlySalesTotal;
use SimpleOrm\Tests\Sample\Models\Role;
use SimpleOrm\Tests\Sample\Models\Transaction;
use SimpleOrm\Tests\Sample\Models\TransactionDetail;
use SimpleOrm\Tests\Sample\Models\User;
use SimpleOrm\Tests\Sample\Models\UserActivityReport;
use SimpleOrm\Tests\Sample\Models\UserProfile;
use SimpleOrm\Tests\Sample\Models\UserRole;
use SimpleOrm\Tests\Sample\Models\UserTransactionTotal;
use SimpleOrm\Tests\Support\ConformancePaths;

/**
 * The conformance runner for entity metadata (CLAUDE.md §9): exports every
 * fixture entity and compares it against `conformance/entities/*.json` — the
 * files every port must reproduce byte-for-byte. CRLF is normalized before
 * comparison because a Windows git checkout can rewrite the pinned files'
 * line endings on disk; the C# reference's own conformance runner does the
 * same for the same reason.
 */
final class EntitiesTest extends TestCase
{
    /** @return array<string, array{0: string, 1: class-string}> */
    public static function entities(): array
    {
        return [
            'user' => ['user', User::class],
            'role' => ['role', Role::class],
            'user_role' => ['user_role', UserRole::class],
            'transaction' => ['transaction', Transaction::class],
            'transaction_detail' => ['transaction_detail', TransactionDetail::class],
            'user_profile' => ['user_profile', UserProfile::class],
            'user_transaction_total' => ['user_transaction_total', UserTransactionTotal::class],
            'monthly_sales_total' => ['monthly_sales_total', MonthlySalesTotal::class],
            'daily_sales' => ['daily_sales', DailySales::class],
            'user_activity_report' => ['user_activity_report', UserActivityReport::class],
        ];
    }

    #[Test]
    #[DataProvider('entities')]
    public function export_matches_the_pinned_conformance_file(string $fileName, string $entityType): void
    {
        $loader = new EntityMapLoader();
        $json = self::normalizeNewlines(EntityMapJson::export($loader->load($entityType), $loader));

        $path = ConformancePaths::dir('entities') . DIRECTORY_SEPARATOR . $fileName . '.json';
        self::assertFileExists($path, "missing conformance file {$path}");
        $expected = rtrim(self::normalizeNewlines((string) file_get_contents($path)), "\n");

        self::assertSame($expected, $json);
    }

    /** Every file conformance/entities/ actually has is covered by a case, so a stray fixture is caught too. */
    #[Test]
    public function every_conformance_entity_file_has_a_case(): void
    {
        $covered = array_map(static fn (array $row): string => $row[0] . '.json', self::entities());
        sort($covered, SORT_STRING);
        self::assertSame(ConformancePaths::cases('entities'), $covered);
    }

    private static function normalizeNewlines(string $text): string
    {
        return str_replace("\r\n", "\n", $text);
    }
}
